[SYCL][PI][L0] Fixes to make sure kernels and programs cannot be destroyed before they have finished execution.

kbsmith-intel opened the pull request:
So, here’s requirements from summary of level zero spec:

1 – cannot do zeKernelDestroy until all commands using the kernel have completed execution as measured by their event or fence being signaled.
2 – The application must destroy all kernel and build log handles created from the module before destroying the module itself (zeModuleDestroy)

So, here is a design I think will work.

1 – We want to do a piKernelRetain sometime before kernel is appended to command-list in piEnqueueKernelLaunch.  The intent is that we will do a piKernelRelease once the Event that we create to be associated with the Enqueued Kernel is signaled/released. The handle of kernel to be released will be stored in the event.

2 – Event currently contains fields for CommandType and CommandData, and for the CommandType that piEnqueueKernelLaunch uses, CommandData is currently unused.  So, for this case, we will set CommandData to be the Kernel, so then when the event is signaled as done, then we can use the kernel we stored in CommandData, we can do piKernelRelease(CommandData).

3 – I think in piKernelRetain, we also need to call piProgramRetain on the program that the kernel is part of, and we need piKernelRelease to call piProgramRelease.  This is necessary to make sure we do not destroy the program before destroying all the kernels that may be in use from the program.

4 – in the destructor for pi_program, we need to destroy the build log before we destroy the program itself.

smaslov-intel commented on sycl/plugins/level_zero/pi_level_zero.cpp:
```diff
@@ -3202,6 +3227,13 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
     (*Event)->Queue = Queue;
     (*Event)->CommandType = PI_COMMAND_TYPE_NDRANGE_KERNEL;
     (*Event)->ZeCommandList = ZeCommandList;
+    (*Event)->CommandData = (void *)Kernel;
```
Add a comment why you save the kernel here

smaslov-intel commented on sycl/plugins/level_zero/pi_level_zero.cpp:
```diff
@@ -572,8 +572,9 @@ bool _pi_queue::isBatchingAllowed() {
-  if (this->isBatchingAllowed()) {
+                                      ze_fence_handle_t ZeFence,
+                                      bool OKToBatchKernel) {
+  if (OKToBatchKernel && this->isBatchingAllowed()) {
```
Instead of adding the new `OKToBatchKernel` argument, could we move this 'if' out of this helper routine to its only caller and do the `executeCommandList` there.

kbsmith-intel replied:
I thought about this, but then the code in piEnqueueKernelLaunch looks like:
if (OKToBatchKernel)
  res = batch
else
 res = execute

I much prefer hiding that inside the batch program, which is what is currently being done by the code.  This also makes it so that other places can use.  An alternative that I considered was moving the batching logic into executeCommandList, and just passing that boolean in there, and deleting batchCommandList entirely.  I would MUCH prefer to do that than to move the OKtoBatchKernel logic out into piEnqueueKernelLaunch.  I could also just change the name of this parameter to OKToBatchThisCommand, and again, either add that to batchCommandList or executeCommandList.

smaslov-intel replied:
I like what you show it would look like in piEnqueueKernelLaunch, this also makes the methods more focused on what they are aimed to do. 

> An alternative that I considered was moving the batching logic into executeCommandList, and just passing that boolean in there, and deleting batchCommandList entirely.

Sounds good too.

kbsmith-intel replied:
Done

smaslov-intel commented on sycl/plugins/level_zero/pi_level_zero.hpp:
```diff
@@ -328,8 +328,10 @@ struct _pi_queue : _pi_object {
   // Attach a command list to this queue and allow it to remain open
   // and used for further batching.  It may be executed immediately,
   // or it may be left open for other future command to be batched into.
+  // OKToBatchKernel indicates whether for this particular kernel in the
```
Even though it is currently only used for `piEnqueueKernelLaunch`, it is created a more general and it is not clear what the "kernel" is here then.

kbsmith-intel replied:
I could change the name to OKToBatchThisCommand?  Would you like it better if I changed this to be a parameter to executeCommandList, defaulted OKToBatchThisCommand tp false, and moved the batching logic to executeCommandList.

smaslov-intel replied:
I think so, yes.

kbsmith-intel replied:
Done

smaslov-intel commented on sycl/plugins/level_zero/pi_level_zero.cpp:
```diff
@@ -2902,6 +2907,10 @@ pi_result piKernelCreate(pi_program Program, const char *KernelName,
   }
+
+  // Update the refcount of the program to show its use by this kernel.
+  piProgramRetain(Program);
```
Just a note: the program refcount will be 2 after piProgramCreate + piKernelCreate, while users are probably expecting 1. I believe OpenCL in this case keeps internal and external reference counting to not mix the two. I don't think we'd need to do this in this change-request though, if at all.

kbsmith-intel replied:
Yes, this might be different than OpenCL.  I don't think we should add an internal ref_count to Program. If we thought that getting ref_count for program to be consistent with OpenCL, then I would simply recommend adding a field to Program, that would be exactly like a ref_count, but only for counting Kernels currently using this program.  That would have a kind of weird effect though.  It would mean that it would be possible to have a Program with a ref count of 0, but that hadn't actually been deleted yet because there were still kernels using it.

smaslov-intel replied:
The objects would get deleted when both user-visible refcount =0 AND internal refcounts all equal to 0 too. Anyways, let's just go with the single one for now.

smaslov-intel commented on sycl/plugins/level_zero/pi_level_zero.cpp:
```diff
@@ -3186,11 +3203,19 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
+  // Should this also force a synchronize immediately?  That would also
+  // ensure that the kernel had completed execution before it could be
+  // released, but that may really hurt performance.
+  bool BatchingThisKernelOK = (Event);
```
At what circumstance would the SYCL RT pass a NULL Event here? Tag @romanovvlad

romanovvlad replied:
Currently SYCL RT always pass non-null pointer to pi_event.

kbsmith-intel replied:
@romanovvlad  - Seems like it would be nice to put in an assert(Event), and take out the if.  Is this true of all the various pi entry points that take Event?

romanovvlad replied:
Agree, if the plugin relies on that it would be helpful to have an assert.

kbsmith-intel replied:
@smaslov-intel - Sergey what do you think about adding an assert and getting rid of the if?

smaslov-intel replied:
Yes, let's do that.

kbsmith-intel replied:
Done.

smaslov-intel commented on sycl/plugins/level_zero/pi_level_zero.cpp:
```diff
@@ -3356,21 +3389,26 @@ pi_result piEventGetProfilingInfo(pi_event Event, pi_profiling_info ParamName,
-  // it being reset to nullptr in another thread.
-  // But, since the ZeCommandList is uniquely associated with the queue
+  // have a race condition between the read of fields of the event,
+  // and resetiing those fields in some other thread.
```
```suggestion
  // and reseting those fields in some other thread.
```

smaslov-intel commented on sycl/plugins/level_zero/pi_level_zero.cpp:
```diff
@@ -3386,6 +3424,13 @@ static void recycleEventCommandList(pi_event Event) {
+  // Release the kernel associated with this event if there is one.
+  if (Event->CommandType == PI_COMMAND_TYPE_NDRANGE_KERNEL &&
+      Event->CommandData) {
+    piKernelRelease((pi_kernel)(Event->CommandData));
+    Event->CommandData = nullptr;
```
```suggestion
    Event->CommandData = nullptr;
    piKernelRelease(pi_cast<pi_kernel>(Event->CommandData));
```

I am not sure if data race is possibly here today, but setting this to null before actually using it seems safer.

kbsmith-intel replied:
The change you suggest is incorrect.  If done the way you suggest, that would just pass nullptr to piKernelRelease always.  There cannot be a race.  The event member Event->CommandData is protected from race by the lock of Queue->PiQueuMutex, the same as Event->ZeCommandList is in the lines just above this.

smaslov-intel replied:
Of course, I meant to save, then nullify, and then operate on the saved one. I see the mutex and it does look sufficent.

smaslov-intel reviewed (changes requested):
LGTM overall, see comments and also:
1) add important design notes to description for history
2) add a test

smaslov-intel commented on sycl/plugins/level_zero/pi_level_zero.cpp:
```diff
@@ -3386,6 +3424,13 @@ static void recycleEventCommandList(pi_event Event) {
+  // Release the kernel associated with this event if there is one.
+  if (Event->CommandType == PI_COMMAND_TYPE_NDRANGE_KERNEL &&
+      Event->CommandData) {
+    piKernelRelease((pi_kernel)(Event->CommandData));
```
```suggestion
    piKernelRelease(pi_cast<pi_kernel>(Event->CommandData));
```

kbsmith-intel commented:
I think I have addressed all your review comments and I have added the basic design notes in the original comment.  I still need to add a test, but please review the revised code while I work on adding a test.  Thank you.
